### PR TITLE
RBAC and Ownable migration towards Roles

### LINKED ISSUE #366: Granularity of permissions
Some contracts need to have an account with special permissions. An example is `MintableToken`, which has an address which is the only one allowed to mint tokens. These are currently implemented as `Ownable`, with the special permissions assigned to the `owner` (by marking some functions as `onlyOwner`).

Sometimes a single contract has more than one feature requiring special permissions. Imagine a `MintableToken` which is also a `PausableToken`. In such a case, the contract owner can both mint tokens and pause the token.

This goes against the security [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege), and also makes some requirements harder to implement. As an example of the latter, suppose we have the requirement to disable minting after a crowdsale, but want to retain the ability to stop the contract in an emergency.

For the two reasons stated above, I believe we need to provide greater granularity of permissions. This is _kind of_ doable under the current ownership framework, by setting as `owner` a contract that can forward calls or not according to some logic. I don't dislike that, but I can see some problems with it.

We might want to provide the granularity _directly_, i.e. by making each privileged feature have a different address (or set of addresses) with permissions for it. At the same time it sounds like it could lead to an explosion of different permissions and roles which could be hard to manage.

Opening up the topic for discussion. Do you agree that permissions need to be more granular? Have you run into this problem yourself?

### LINKED ISSUE #1090: Improve RBAC role type
As of v1.11.0, we're using [strings](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/release-v1.11.0/contracts/ownership/rbac/RBAC.sol#L20) as roles in RBAC. This can be problematic, because strings are [UTF-8 encoded](https://solidity.readthedocs.io/en/v0.4.24/types.html#dynamically-sized-byte-array) in Solidity. Therefore, the following is valid Solidity code, and looks correct at a first glance:

```
contract Campaign is RBAC {
  function () external payable {
    require(msg.value > 1 ether);
    addRole(msg.sender, "backer");
  }

  function doBackerStuff() public onlyRole("backer​") {
    ...
  }
}
```

`doBackerStuff`'s modifier, however, does not contain the original string: an invisible [ZERO WIDTH SPACE (U+200B)](https://www.fileformat.info/info/unicode/char/200b/index.htm) character has been added at the end, causing seemingly correct code to contain an (intentional?) bug.

A suggested alternative is to use integers as role identifiers, which removes this issue entirely. Assigning new identifiers to a role becomes a highly arbitrary task however, since numbers don't convey any role information, as opposed to strings like 'admin', 'curator', 'minter', etc. This could be sidestepped by using the hash of the role string as the integer identifier, e.g., `uint256 public constant adminRole = 0x01ffc9a7...; // 0x01ffc9a... == keccak256('admin')`, at the cost of placing the extra burden of having to use this pattern on the library users.

### LINKED ISSUE #1146: Migrate Ownable to Roles
With the 2.0 release coming soon, we want to stabilize as much of our API as possible, and in the process reduce the attack surface of our contracts. It's been pointed out that `RBACOwnable` is very similar to `Ownable` and could easily replace it, providing future extensibility, at the cost of added complexity.

We need to decide whether to keep both contracts or remove `Ownable` altogether from the codebase (which would impact some contracts like `Heritable`). I personally think this duplication is a mistake, which has caused the `owner` concept to extend to places where it doesn't belong (e.g. `Mintable`), and that we'd benefit greatly from the role semantics (`minter`, `burner`, etc.). This would also mean dropping the `RBAC` prefix from multiple contracts, since that's our standard role solution (e.g. `RBACOwnable` and `RBACMintable` would simply be `Ownable` and `Mintable`).

This is a place to discuss whether or not we want to do this: if consensus is reached and we decide to move forward, we'll also need to settle on a design for the affected contracts (like `Inheritable`).

Resolves https://github.com/OpenZeppelin/zeppelin-solidity/issues/366.

### PR DESCRIPTION
**Pull request by nventuro**
Notable changes:
 * `Ownable` is no longer used (except in drafts)
 * `RBAC` has been removed
 * `Secondary` was introduced to replace `Ownable` in simple cases (e.g. `Escrow`)
 * The `Roles` library is now much more widely used, though the role contracts themselves have to be manually generated (see #1290)

Closes #366
Closes #1090
Closes #1146

### CODE REVIEW

**Comment by frangio** [comment]
On hold for #1276.

File: contracts/token/ERC20/ERC20Detailed.sol [hunk 1]
```diff
@@ -10,34 +10,13 @@ import "./IERC20.sol";
  * just as on Ethereum all the operations are done in wei.
  */
 contract ERC20Detailed is IERC20 {
-  string private name_;
-  string private symbol_;
-  uint8 private decimals_;
+  string public name;
+  string public symbol;
+  uint8 public decimals;
```
Comment (elopio) [inline contracts/token/ERC20/ERC20Detailed.sol]: hey, what happened here? I suppose this was a failed merge.

File: contracts/token/ERC721/ERC721Mintable.sol [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * @title ERC721Mintable
+ * @dev ERC721 minting logic
+ */
+contract ERC721Mintable is ERC721, MinterRole {
+  event Minted(address indexed to, uint256 tokenId);
+  event MintingFinished();
+
+  bool public mintingFinished = false;
```
Comment (elopio) [inline contracts/token/ERC721/ERC721Mintable.sol]: public? Nooo

File: contracts/access/roles/CapperRole.sol [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  modifier onlyCapper() {
+    require(isCapper(msg.sender));
+    _;
+  }
+
+  function isCapper(address _account) public view returns (bool) {
+    return cappers.has(_account);
+  }
+
+  function addCapper(address _account) public onlyCapper {
```
[the remark this answers is not in the document]
Comment (shrugs) [reply contracts/access/roles/CapperRole.sol]: that's a good point; a lot of the api surface for these runs into the same issues with the problem with RBAC and it's that you can't define a public api surface by default because then changing it is super hard. and you don't know what sort of process a project might want to be able to use to be able to change the cap, for example.

File: contracts/token/ERC20/ERC20Detailed.sol (hunk 1 above)
Comment (shrugs) [reply contracts/token/ERC20/ERC20Detailed.sol]: I think this needs to be merged into the new master again? not sure

File: contracts/token/ERC721/ERC721Mintable.sol (hunk 2 above)
Comment (shrugs) [reply contracts/token/ERC721/ERC721Mintable.sol]: good point; if we update this, also update the `ERC20Mintable` that this was copy/pasted from

File: contracts/token/ERC20/ERC20Detailed.sol (hunk 1 above)
Comment (frangio) [reply contracts/token/ERC20/ERC20Detailed.sol]: We must've merged this file wrong.

File: contracts/access/roles/CapperRole.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (frangio) [reply contracts/access/roles/CapperRole.sol]: Our rationale was that someone who has a role is already able to perform actions using that role on behalf of other people. In fact, with ownership, it was trivial to transfer ownership to a forwarder contract that would let any set of accounts use the ownership. But this could also be done informally, as an off-chain favor.

Since this was already possible, we decided to make it part of the API. Otherwise it just feels like obscuring what is really possible.

[the remark this answers is not in the document]
Comment (frangio) [reply contracts/access/roles/CapperRole.sol]: Our rationale was that someone who has a role is already able to perform actions using that role on behalf of other people. In fact, with ownership, it was trivial to transfer ownership to a forwarder contract that would let any set of accounts use the ownership. But this could also be done informally, as an off-chain favor.

Since this was already possible, we decided to make it part of the API. Otherwise it just feels like obscuring what is really possible.

File: contracts/access/Roles.sol [hunk 4]
```diff
@@ -15,32 +13,17 @@ library Roles {
   /**
    * @dev give an account access to this role
    */
-  function add(Role storage _role, address _account)
-    internal
-  {
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/access/Roles.sol]: Not sure what you mean D:

File: contracts/access/roles/CapperRole.sol [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    cappers.add(msg.sender);
+  }
+
+  modifier onlyCapper() {
+    require(isCapper(msg.sender));
+    _;
+  }
+
+  function isCapper(address _account) public view returns (bool) {
+    return cappers.has(_account);
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/access/roles/CapperRole.sol]: Agree on all points, including this being improvable later on.

File: contracts/access/roles/CapperRole.sol [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  function addCapper(address _account) public onlyCapper {
+    cappers.add(_account);
+  }
+
+  function renounceCapper() public {
+    cappers.remove(msg.sender);
+  }
+
+  function _removeCapper(address _account) internal {
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/access/roles/CapperRole.sol]: We've discussed this extensively with @frangio. Our reasoning goes as follows:

An account with a role should always be able to transfer it. If the account is a contract, then the conditions under which a transfer will be executed will be clear from the code. If it's an EOA, then the user already has the liberty to do whatever she wants with it, from complying to transaction requests to giving up the private keys to someone else, effectively transferring control.

So, transferring can always be done. This includes transferring the role to a contract that will forward all calls from a whitelisted set of addresses, which effectively is an `add` to multiple addresses. We decided to formalize this mechanism, and simply remove `transfer` and make `add` public by default. Note that `remove` plays no part here: at no point does having a role give you the power to take it away from someone else, but it does let you act at someone else's behest.

TL;DR: EOAs shouldn't be trusted, since people can do anything they want, if you _really_ need control over this kind of stuff, the role-havers should also be contracts.

File: contracts/access/roles/MinterRole.sol [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+contract MinterRole {
+  using Roles for Roles.Role;
+
+  Roles.Role private minters;
+
+  constructor() public {
+    minters.add(msg.sender);
+  }
+
+  modifier onlyMinter() {
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/access/roles/MinterRole.sol]: I probably should've gone into some more detail in the PR description: the role contracts (everything under `access/roles/`, and associated tests) will be auto-generated in the future (see #1290), since they are all identical (we call it the 'role boilerplate'). This is sort of a limitation of how libraries work in Solidty, but I'm not yet convinced what'd be the best way to tackle this moving forward. For now, code repetition is what we have.

**Comment by nventuro** [comment]
Thanks for the thorough review @elopio @shrugs @frangio :tada:

**Comment by mjdietzx** [comment]
Just to be clear, the recommended way to implement `Whitelist` functionality after this change is to do something like:

- Create a `Whitelist` contract using `Roles` similar to https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/access/roles/CapperRole.sol
- Implement something like `addBatch` and `removeBatch`
- Check against the whitelist as normal (ie in `MyCrowdSale`)

Just so I can understand, why did you decide to remove the whitelist and whitelisted crowd-sale? Was it just not adding core functionality / unnecessary? Or is there some deeper reason?

**Comment by nventuro** [comment]
Hey @mjdietzx! There's a short explanation [here](https://github.com/OpenZeppelin/openzeppelin-solidity/issues/1292): the gist of it is that, if implemented using the current role system, the whitelistees would also become whitelisters themselves. This may or may not be an issue: we need to figure out what our purpose is with those contracts, and them implement them. Since 2.0 now features a [stable API](https://github.com/OpenZeppelin/openzeppelin-solidity/releases/tag/v2.0.0-rc.1), I didn't want to include a contract we weren't 100% sure about.

**Comment by mjdietzx** [comment]
Hey @nventuro, thanks for the response! Just some feedback, I think `Whitelist.sol` and whitelisted crowd-sales in general are a very common use-case (kyc & aml keep getting stricter 😕) And a whitelisted address being able to add another address to the whitelist wouldn't be acceptable for this use-case. So I would think it's important to be able to implement this well in openzeppelin.

Either way, we'll be updating to v2.0 (looking forward to it) and will find a good solution for our whitelist with whatever you decide to go with. Looking forward to the release, thanks!

**Comment by frangio** [comment]
@mjdietzx Please check out my comment in https://github.com/OpenZeppelin/openzeppelin-solidity/issues/1292 and let us know what you think over there!
